Added support `F.one_hot`

shaharelys opened the pull request:
<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs? No, should I?
- [x] Did you write any new necessary tests?

</details>

## What does this PR do?
Implements `F.one_hot`.

Fixes #64 

## PR review
This PR is open for review, yet again I'm not certain it's complete. Added comments under `one_hot` for things I wasn't sure of. Tests have passed. `docs` has not been updated (I had a look at this [README](https://github.com/Lightning-AI/pytorch-lightning/blob/master/docs/README.md) but I wasn't sure I understand). Feedback is welcomed!

## Did you have fun?
I sure did! 🙃

mruberry commented on thunder/tests/opinfos.py:
```diff
@@ -6152,6 +6152,32 @@ def sample_generator(op, device, dtype, requires_grad, **kwargs):
+    max_value = 9
+    for shape in test_shapes:
+        for num_classes in range(1, max_value + 1):
+            a = make(shape, low=0, high=num_classes - 1)  # use non-negative integers
```
Nice comment

mruberry commented on thunder/tests/opinfos.py:
```diff
@@ -6152,6 +6152,32 @@ def sample_generator(op, device, dtype, requires_grad, **kwargs):
 
 
+def one_hot_sample_generator(op, device, dtype, requires_grad, **kwargs):
+    make = partial(make_tensor, device=device, dtype=torch.long, requires_grad=requires_grad)
```
`dtype=dtype` and let's restrict the generation dtype to be `torch.long` elsewhere

mruberry commented on thunder/tests/opinfos.py:
```diff
@@ -6152,6 +6152,32 @@ def sample_generator(op, device, dtype, requires_grad, **kwargs):
+    ltorch.one_hot,
+    sample_input_generator=one_hot_sample_generator,
+    torch_reference=torch.nn.functional.one_hot,
+    dtypes=(datatypes.signedinteger,),
```
Does PyTorch actually support dtypes other than int64? In the generator the dtype argument is ignored, so all the generated tests will run with int64 tensors, even though they claim to be testing other signed integer types.

mruberry replied:
Reviewing the rest of the PR, maybe this should only generate tests for additional signed integers when using the nvFuser executor? If that's the case then dtypes other than int64 can be skipped when using the torch executor. 

Here's an example of an executor-specific test directive:

https://github.com/shaharelys/lightning-thunder/blob/9ff2c62491e52d580d8acfead46e15b40773fa1e/thunder/tests/opinfos.py#L789

shaharelys replied:
Not sure this is resolved.

mruberry commented on thunder/torch/__init__.py:
```diff
@@ -3489,6 +3489,21 @@ def embedding_backward(grad, indices, num_weights, padding_idx, scale_grad_by_fr
 
+@torchsymbol(torch.nn.functional.one_hot, id="torch.nn.functional.one_hot", is_method=False)
+def one_hot(a: TensorProxy, /, num_classes: int) -> TensorProxy:  # Should we use TensorProxy for a?
+    if num_classes == -1:
```
I think this can be simplifiedto `utils.check(num_classes >= 0, ...)`

mruberry commented on thunder/torch/__init__.py:
```diff
@@ -3489,6 +3489,21 @@ def embedding_backward(grad, indices, num_weights, padding_idx, scale_grad_by_fr
 
 
+@torchsymbol(torch.nn.functional.one_hot, id="torch.nn.functional.one_hot", is_method=False)
+def one_hot(a: TensorProxy, /, num_classes: int) -> TensorProxy:  # Should we use TensorProxy for a?
```
To be consistent with other type annotations in this file the annotation would be:

```
a: TensorLike, /, num_classes: int) -> TensorLike:
```

mruberry commented on thunder/torch/__init__.py:
```diff
@@ -3489,6 +3489,21 @@ def embedding_backward(grad, indices, num_weights, padding_idx, scale_grad_by_fr
+    # TODO: should we implement this check? I suspect this might hinder performance, also we do not support .any)
+    #  utils.check(a.any() >= 0, lambda f"input tensor should have non-negative values", exception_type=ValueError)
+
+    canvas = zeros(*a.shape, num_classes, dtype=dtypes.int64)  # this is equivalent to the use of .long in torch
```
The test failures are because this tensor is always allocated on the cpu. 

The best way to fix this (which doesn't need to happen in this PR) is to extend the "like" functions, like `zeros_like`

https://github.com/shaharelys/lightning-thunder/blob/9ff2c62491e52d580d8acfead46e15b40773fa1e/thunder/torch/__init__.py#L526

to accept a `shape` parameter that can override the shape of the example tensor. Then the line could be:

```
result_shape: tuple[int, ...] = a.shape + (num_classes,)
canvas = zeros_like(a, shape=result_shape)
```

and developers wouldn't need to worry about setting additional metadata.

For now, the simplest fix is to just set the device:

```
canvas = zeros(*a.shape, num_classes, device=a.device, dtype=dtypes.int64)
```

fyi @nikitaved to consider the extension of the "like" operations

nikitaved replied:
We have `zeros_like`

mruberry replied:
We have `zeros_like`, but I don't think it has a shape parameter to override the example tensor's shape?

shaharelys replied:
Was not sure about this one

mruberry commented on thunder/torch/__init__.py:
```diff
@@ -3489,6 +3489,21 @@ def embedding_backward(grad, indices, num_weights, padding_idx, scale_grad_by_fr
+
+    canvas = zeros(*a.shape, num_classes, dtype=dtypes.int64)  # this is equivalent to the use of .long in torch
+    index = a.unsqueeze(-1)
+    src = ones_like(index, dtype=dtypes.int64)
```
Like the above `canvas` tensor, this tensor is also always allocated on the cpu

mruberry replied:
The construction of `src` avoids the above issue but requires unsqueezing `a` just to use that result as an example. I suggest mimicking the above approach and calling `ones`, and computing the desired shape as `a.shape + (1,)`.

mruberry commented on thunder/tests/opinfos.py:
```diff
@@ -6152,6 +6152,32 @@ def sample_generator(op, device, dtype, requires_grad, **kwargs):
+    make = partial(make_tensor, device=device, dtype=torch.long, requires_grad=requires_grad)
+
+    test_shapes = [
+        (10,),
```
Let's add a tensor with no dimensions (its shape is `()`) and a tensor with no elements (like `(0, 512)`, too

nikitaved replied:
Yeah, empty tensors + 0-dim tensors are very important. The underlying ops dispatch to PyTorch native implementations but these were written by me before the widespread adoption of 0-dim tensors... So we'd better double check that PyTorch as well does the right thing here...

shaharelys replied:
@nikitaved I've added an empty input `(, 512)` as you guys suggested (:

nikitaved replied:
@shaharelys , what about scalar inputs, the inputs with no dimension?

shaharelys replied:
Hey @nikitaved! Sorry for the delay. I did not add these. Should we also add these now?

mruberry commented:
Hey @shaharelys! This looks pretty good. I made some comments for your review. I also added @nikitaved as a reviewer.

shaharelys commented:
Thx a lot! Will look into these 🙏🏼

nikitaved commented on thunder/torch/__init__.py:
```diff
@@ -3489,6 +3489,21 @@ def embedding_backward(grad, indices, num_weights, padding_idx, scale_grad_by_fr
+
+    canvas = zeros(*a.shape, num_classes, dtype=dtypes.int64)  # this is equivalent to the use of .long in torch
+    index = a.unsqueeze(-1)
+    src = ones_like(index, dtype=dtypes.int64)
+
+    return scatter_add(canvas, dim=-1, index=index, src=src)
```
Seems fine for now, but what happens is that we create a tensor full of ones. We could:
- create a tensor with a single 1 and broadcast it.
- leave it be in hopes that sometime in the future our backends/interpreters will improve this bit (when scatter_add will be there for nvFuser, for example). Worth adding a comment, I guess?

shaharelys replied:
Hey @nikitaved! Not sure I understand the implementation suggested in _..and broadcast it_. Tried to implement naively by replacing,
'src = ones_like(index, device=a.device, dtype=dtypes.int64)'
with,
`src = tensor([1], device=a.device, dtype=dtypes.int64)` 
and got,
```
thunder/core/prims.py:3009: in scatter_add_meta
    utils.check(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cond = False, s = <function scatter_add_meta.<locals>.<lambda> at 0x7bb2a8c043a0>
exception_type = <class 'RuntimeError'>

    def check(cond: bool, s: Callable[[], str], exception_type: type[Exception] = RuntimeError) -> None:
        """Helper function for raising an error_type (default: RuntimeError) if a boolean condition fails.
    
        s is a callable producing a string to avoid string construction if the error check is passed.
        """
        if not cond:
>           raise exception_type(s())
E           RuntimeError: Expected index (rank=3) to have the same rank as value (rank=1)

thunder/core/baseutils.py:103: RuntimeError
```

nikitaved replied:
You can read about broadcasting here: https://numpy.org/doc/stable/user/basics.broadcasting.html

shaharelys replied:
Cool, and should this be more efficient operation than current implementation?

nikitaved replied:
Well, yes. It will spare us launching a kernel that fills memory with `1`s. This might become redundant once the target executor can fuse together `ones` and `scatter_add`. In the context of NVFuser, this means putting the code of `ones` and `scatter_add` into a single CUDA kernel. Do not worry for now if you do not understand these things, this could be done in some follow-up. You can learn more about executors from the documentation that should be, alas, built locally.

shaharelys commented:
Hey @mruberry, @nikitaved ! I've reviewed and addressed most of the comments. However, there are a few points I'm uncertain about. I'll comment directly under those for clarity.

Looking forward to your feedback!

shaharelys commented on thunder/torch/__init__.py:
```diff
@@ -3489,6 +3489,21 @@ def embedding_backward(grad, indices, num_weights, padding_idx, scale_grad_by_fr
+        # TODO: implement when we're ready to support auto inference using `.item()`
+        utils.check(False, lambda: "num_classes=-1 not supported yet.", exception_type=NotImplementedError)
+    # TODO: should we implement this check? I suspect this might hinder performance, also we do not support .any)
+    #  utils.check(a.any() >= 0, lambda f"input tensor should have non-negative values", exception_type=ValueError)
```
Was not sure about this one. Just revised the `# TODO..` for now.

nikitaved replied:
My message could be interpreted as that we do not have to do anything, as long as `scatter`-like ops are PyTorch primitives. Not sure about NVFuser though...

t-vi reviewed (approved):
Thank you @shaharelys @nikitaved @mruberry

nikitaved commented:
@t-vi , I have not checked the charges yet, nor have I answered the posited questions :)
